### ISSUE TITLE
"fi_worker_mp" cannot use more than 200% cpu usage

### ISSUE THREAD

**Issue by timroland** [amusecode/amuse#760]
Hello !

**Describe the bug**
When using Fi in "openmp" mode, I cannot get more than 200% CPU usage dedicated to the "fi_worker_mp" process, whatever the environment variable OMP_NUM_THREADS is set to. Openmp does not seem to share correctly the computation load: when running htop, I can see a correct number of threads running (i.e. corresponding to the OMP_NUM_THREADS value), however most of them only use a fraction of CPU power (see screenshot below).

**Expected behavior**
Expect openmp to spread the work load to the number of threads specified by OMP_NUM_THREADS (to scale up my simulations !)

**Logs**
![Fi_openmp](https://user-images.githubusercontent.com/32935374/116440733-10a50680-a851-11eb-8443-af33a6f497ab.png)
In the log file of Fi, the OMP_NUM_THREADS value is also well retrieved.
[Fi.log](https://github.com/amusecode/amuse/files/6393492/Fi.log)

**Environment (please complete the following information):**
I have the same problem on 2 different machines (Amuse is built the same way on both): 
A computing server (80 cores):
 - OS and version: Debian 10
 - Compiler: gcc8.3

and a basic workstation:
 - OS and version: Ubuntu 20.04
 - Compiler: gcc9.3

**Additional context**
Note that on a laptop with Ubuntu 16.04 and gcc7.5 (but only 4 cores), everything works fine. Would it be only a problem of gcc version ?

**Comment by timroland** [comment]
Hi,

We made some investigations out of the AMUSE framework, to test openmp on our server with other codes directly written in fortran, and it seems to work fine, discarding the version issue. We think that the problem would rather reside in the interface between python and fortran. I did not understand how it works exactly but it seems that python launch the processes only on a single core. However, the server is made with only two threads per core, so we would like to use multiple cores to exploit all the threads of the machine (there are 40 cores with 2 threads per core). Would it be possible to overcome this limitation ? 

Thanks again for your help !

**Comment by ipelupessy** [comment]
if its a shared memory machine (e.g. if you do "top" all 80 virtual cpus show up), its a matter of configuring the OpenMP environment, e.g.:
```export OMP_NUM_THREADS=80```
and then start your script, though usually this is not necessary...

**Comment by timroland** [comment]
Yes, it is a shared memory machine. We have already try to set different value of the OMP_NUM_THREADS variable but nothing changed... What surprises me the most is that in the logfile of Fi, the value of OMP_NUM_THREADS is well retrieved when the code is launched.  So it must be something else to configure.

**Comment by rieder** [comment]
Do you have any other (OMP) environment variables set? You’re not running via e.g. SLURM?

**Comment by ipelupessy** [comment]
hmm..it seems on the screen shot almost like the openmp restricts itself to one or two cores/cpus...are you using the repository or an installed amuse via pypi? you might try to set the environment OMP_DYNAMIC to false, this is done in line 107 of mem.f90:
```!$ call omp_set_dynamic(.FALSE.)```
and swapping this to .TRUE. might also do something (the former you can do if you have the pypi install the latter if you are compiling from source...

**Comment by ipelupessy** [comment]
so @rieder suggestion could also be the problem (that a workload manager is restricting the process to some cpu...)

**Comment by timroland** [comment]
@rieder: I don't have (to my knowledge) other OMP environment variables set and I run my script directly with python (as you can see on the screenshot, PID 9080), I sometimes use mpirun or mpiexec but didn't notice any difference or advantages (maybe am I wrong ?) 

@ipelupessy: I am compiling amuse from source, in a specific virtual environmment (I compile it in "development mode", so with pip install -e .). I suppose that I have to recompile the code, after I made your modification in mem.f90, right ? Anyway I gave it a try and I did not see any change.

**Comment by ipelupessy** [comment]
hmm..can you compile and run the following program:
```
program test
 integer omp_get_num_procs, omp_get_max_threads

 print*, omp_get_num_procs(),omp_get_max_threads()

end program
```
(putting it in a test.f90 file and then gfortran -fopenmp test.f90)..and post the result?

**Comment by lgautier4** [comment]
Hello, i try to run fi  with timroland. 
OMP is working well, all core are used if we use a program directly with fortran. 
We use the source code and compiling.  The job scheduler is not the problem, all cores can be used, and i tried also the code in a local machine. If call omp_set_dynamic(.true.) then, just 2 process are running  whatever OMP_NUM_THREAD equal is.
If it is false the number process asked is running but in one core. 

The problem: if we use python to run fi using OMP. All threads asking via OMP_NUM_THREAD are running but only in one core and are not distributed in other cores. Im wondering if its due to gil of python?


I would like to thank you, most sincerely for your support

**Comment by ipelupessy** [comment]
ok, its not the gil of python (the worker process of fi is running as a completely seperate executable, which does not even know its started from python)..
so, setting omp_set_dynamic(.true.) enables dynamic allocation of threads (enabling should not matter, but its good to test because I thought maybe it had something do to with that)..in any case you can leave it the default (.false.) now..the behaviour does seem to indicate that the executable after its start through the mpi spawn is restricted to use just 1 core. Now, I don't think this behaviour is standard, though often MPI is configured a bit restrictive

can you try to run another openmp parallel (test) code and see what it does if you start it like:
mpiexec -n 1 <executable>
?? 
will try to see if this behaviour is documented somewhere..

**Comment by ipelupessy** [comment]
btw, is there a hostfile that is being used/ can be picked up?

**Comment by ipelupessy** [comment]
e.g. a "openmpi-default-hostfile" somewhere in /etc or /usr

**Comment by lgautier4** [comment]
Ok, thank you,  so its a openmpi problem. 
I tried with the option " --map-by socket:pe n"  where n is the number of core by cpu:
 mpirun -n 1 --map-by socket:pe n  PythonCode
It seem to work.
Im not familiar with this option, i have to test more precisely the option. I will tell you if all is ok.

If i run an openmp code with  "mpiexec -n 1 " , the code work with just one core and the number of thread equal to OMP_NUM_THREADS  work in one core.
All is commented in /etc/openmpi-default-hostfile

**Comment by ipelupessy** [comment]
ok, I can reproduce your problems with newer openmpi4 and indeed confirm it is some binding issue, specifically:
if i run the following script:
```
from amuse.ic.gasplummer import new_plummer_gas_model
from amuse.units import nbody_system
from amuse.community.fi.interface import Fi
gas=new_plummer_gas_model(10000)
f=Fi(mode="openmp", redirection="none")
f.gas_particles.add_particles(gas)
f.evolve_model(1. | nbody_system.time)
print(f.model_time)
```
with: ```mpirun -n 1 --report-bindings python fi_minimal.py``` I get on my 4 core laptop:
```
[..] MCW rank 0 bound to socket 0[core 0[hwt 0-1]]: [BB/..]
[..] MCW rank 0 bound to socket 0[core 1[hwt 0-1]]: [../BB]
  --- FI Amuse interface ---
  omp threads:           2
1.0 time
```
this is different from openmpi2 where it simply uses 4 threads. This is not so bad in the sense that it tries to seperate the different processes, but ofcourse you would want more cores assigned to the worker...

Now if I try to change the binding policy: ```mpirun -n 1 --report-bindings --bind-to none  python fitest.py```
it reports:
```
[..]MCW rank 0 is not bound (or bound to all available processors)
[..]MCW rank 0 bound to socket 0[core 0[hwt 0-1]]: [BB/..]
```
so it doesn't bind the main process, but the spawn is still the same (and it uses 2 threads)..now I change the spawn behaviour, but up to now I have only managed this by changing the channel code, adding a keyword to the MPI_Info passed to spawn:
```
        self.info = MPI.Info.Create()
        self.info['bind_to'] = "none"
 ```
 in the ```__init__``` of line 882 in channel.py, in this case:
 ```
[..] MCW rank 0 is not bound (or bound to all available processors)
[..] MCW rank 0 is not bound (or bound to all available processors)
  --- FI Amuse interface ---
  omp threads:           4
```
We should maybe consider this issue carefully a bit more, because I am not sure which behaviour is best (ideally there would be some way to specify the number of cores to use..probably there is..)

**Comment by ipelupessy** [comment]
It may warrant a new issue to develop some concept of managment of resources in case of openmp worker..

**Comment by lgautier4** [comment]
ok, i tried it worked well.
in my case (biprocesseur with 20 cores):
mpirun -n 1 --map-by numa:pe=20 PythonCode work in one socket. It work faster than --bind-to none option (mpirun -n 1 --map-by numa:pe=20 --bind-to none PythonCode) which work for the 2 sockets.

**Comment by ipelupessy** [comment]
its mostly fixed, but I think keep open for a while..
